File: conformance-tests/fixtures/canonical-json/parity_test.py
"""
Canonical JSON parity test — Python

Loads the shared vector pack and verifies that the Python SDK's canonical_json()
produces identical output for every test vector.
"""

import json
import sys
from pathlib import Path

# Add the Python SDK to the import path
sdk_path = Path(__file__).resolve().parents[3] / "reference-implementation" / "python" / "src"
sys.path.insert(0, str(sdk_path))

from osp.manifest import canonical_json  # noqa: E402


def load_vectors() -> list[dict]:
    vectors_path = Path(__file__).parent / "vectors.json"
    with open(vectors_path) as f:
        data = json.load(f)
    return data["vectors"]


def test_canonical_json_parity():
    """Verify Python canonical_json matches every shared test vector."""
    vectors = load_vectors()
    passed = 0
    failed = 0

    for v in vectors:
        actual = canonical_json(v["input"])
        if actual == v["expected"]:
            passed += 1
        else:
            failed += 1
            print(f"FAIL [{v['id']}]: {v['description']}")
            print(f"  expected: {v['expected']}")
            print(f"  actual:   {actual}")

    print(f"\nCanonical JSON parity (Python): {passed}/{len(vectors)} passed")
    return failed == 0


if __name__ == "__main__":
    success = test_canonical_json_parity()
    sys.exit(0 if success else 1)

File: conformance-tests/fixtures/canonical-json/parity-test.ts
/**
 * Canonical JSON parity test — TypeScript
 *
 * Loads the shared vector pack and verifies that the TS SDK's canonicalJson()
 * produces identical output for every test vector.
 */

import { readFileSync } from "node:fs";
import { join } from "node:path";
import { canonicalJson } from "../../reference-implementation/typescript/src/crypto.js";

interface Vector {
  id: string;
  description: string;
  input: unknown;
  expected: string;
}

interface VectorPack {
  vectors: Vector[];
}

const pack: VectorPack = JSON.parse(
  readFileSync(join(__dirname, "vectors.json"), "utf-8"),
);

let passed = 0;
let failed = 0;

for (const v of pack.vectors) {
  const actual = canonicalJson(v.input);
  if (actual === v.expected) {
    passed++;
  } else {
    failed++;
    console.error(`FAIL [${v.id}]: ${v.description}`);
    console.error(`  expected: ${v.expected}`);
    console.error(`  actual:   ${actual}`);
  }
}

console.log(`\nCanonical JSON parity (TypeScript): ${passed}/${pack.vectors.length} passed`);
if (failed > 0) {
  process.exit(1);
}
